Update to NgModule Registration

alxhub opened the pull request:
This PR refactors and updates Angular's NgModule registration scheme, which supports lazy loading in bundling environments where direct references are not available to dynamically loaded ES modules, and global side effects must be used to rendezvous with loaded code.

In particular, this PR addresses an issue where previously this system relied upon such side effects in `.ngfactory` shim files, which are deprecated and in the process of being removed.

See individual commits for more details.

alxhub committed:
perf(compiler-cli): ignore the module.id anti-pattern for NgModule ids

In early versions of Angular, it was sometimes necessary to provide a
`moduleId` to `@Component` metadata, and the common pattern for doing this
was to set `moduleId: module.id`. This relied on the bundler to fill in a
value for `module.id`.

However, due to the superficial similarity between `Component.moduleId` and
`NgModule.id`, many users ended up setting `id: module.id` in their
NgModules. This is an anti-pattern that has a few negative effects,
including preventing the NgModule from tree-shaking properly.

This commit changes the compiler to ignore `id: module.id` in NgModules, and
instead provide a warning which suggests removing the line entirely.

alxhub committed:
refactor(compiler-cli): support non-error diagnostics from traits

Previously, the `TraitCompiler` would naively consider a compilation as
failed if either analysis or resolution produced any diagnostics. This
commit adjusts the logic to only consider error diagnostics, which allows
warnings to be produced from `DecoratorHandler`s.

This is a precursor commit to introducing such a warning. As such, the
logic here will be tested in the next commit.

AndrewKushnir reviewed (approved):
Looks great, thanks @alxhub 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

atscott commented on aio/content/errors/NG10003.md:
```diff
@@ -0,0 +1,22 @@
+
+Doing so is not a common practice, and is only useful in very specific bundling scenarios. Declaring an `id` also makes NgModules non-tree-shakeable, so it should only be done if necessary.
+
+In earlier versions of Angular, it was sometimes necessary to include the property `moduleId: module.id` in `@Component` metadata. Using `module.id` for NgModules is a misunderstanding of the purpose of `id`.
```
A note on the commit message: I think it would be useful to expand it even more. In my opinion this isn't quite enough given what I know from the previous commit: "This is an anti-pattern that has a few negative effects, including preventing the NgModule from tree-shaking properly." The fact that it can't tree shake properly is simply a side effect of using `id` at all in the `NgModule`, not `module.id`. It's still not clear to me why `module.id` specifically isn't useful.

This might apply to this part of the documentation as well. It kind of reads like "`id: module.id` doesn't make sense. Just trust me."

I _think_ I can extrapolate why `module.id` isn't useful: `id` is used for lazy loading and looking up a module by that string id. But you likely can't do that effectively if you're using `module.id` because you don't really know what that value is from another file so you should really just use a known static value. Is that right?

alxhub replied:
I've expanded the error guide as suggested!

JoostK commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -1244,10 +1244,30 @@ function allTests(os: string) {
+      expect(getDiagnosticSourceCode(diags[0])).toEqual('module.id');
+    });
+
+    it('should emit a side-effectful registration call when an @NgModule has an id', () => {
```
A compliance test would be best here, as that performs a roundtrip in partial compilation mode.

That said: this effectively means that `ɵɵregisterNgModuleType` has to be treated as stable API, as partial compilation will also include this call as a top-level statement. Are we okay with that?

alxhub replied:
(as discussed offline) this doesn't actually affect partial compilation, as the `ɵɵregisterNgModuleType` side effect is only added in the full `compileNgModule` operation, not in the `compileDeclareNgModuleFromMetadata` partial compilation.

atscott reviewed (approved):
reviewed-for: public-api

alxhub committed:
refactor(compiler-cli): linker honors associated statements for a field

When `@angular/compiler` processes metadata and compiles a definition field,
it might also choose to return statements that are associated with that
definition, and should be included after the type being compiled. Currently,
the linker ignores these statements, as there are none generated that are
relevant in the linking operation.

A challenge to supporting such associated statements is that the linker
operates on "declare" expressions, and replaces those expressions with other
expressions. It does not have the capability to append statements after the
whole type. The linker actually faces this challenge with statements from
the `ConstantPool` as well, and solves this problem by generating an IIFE
expression that executes the statements and then returns the definition
expression.

Previously, an `EmitScope` processed the definition and converted it to an
expression, as well as collected constant statements from a `ConstantPool`.
A special `IifeEmitScope` implementation was used when emitting into a
context where top-level constant statements couldn't be added at all, and
uses the IIFE strategy in this case.

This commit adds blanket support for associated statements to the linker
using this IIFE strategy. The main `EmitScope` now uses the IIFE strategy to
emit associated statements, and `IifeEmitScope` has been renamed to
`LocalEmitScope`. Now, the `LocalEmitScope` represents constant statements
as associated statements to the main `EmitScope` implementation, so they
get included in the IIFE as well.

Tests are adjusted/added to cover this new behavior. This is a refactoring
commit because no live generated code is affected - there are no cases where
associated statements are present in linked definitions today.

alxhub commented:
Marking this as blocked since the partial compilation tests need a bit of work still.

JoostK commented on packages/compiler-cli/linker/src/file_linker/emit_scopes/emit_scope.ts:
```diff
@@ -23,16 +26,33 @@ export class EmitScope<TStatement, TExpression> {
+      // insert statements after definitions. To work around this, the linker transforms the
+      // definition into an IIFE which executes the definition statements before returning the
+      // definition expression.
+      const importGenerator = new LinkerImportGenerator(this.ngImport);
```
Consider pulling this up because it's also used earlier

JoostK commented on packages/compiler-cli/linker/src/file_linker/emit_scopes/local_emit_scope.ts:
```diff
@@ -0,0 +1,40 @@
+   * are already emitted in the IIFE alongside the translated definition.
+   */
+  override getConstantStatements(): TStatement[] {
+    throw new Error('BUG - IifeEmitScope should not expose any constant statements');
```
```suggestion
    throw new Error('BUG - LocalEmitScope should not expose any constant statements');
```

alxhub committed:
refactor(compiler): change NgModule scoping emit flag to enum

The `compileNgModule` operation previously supported a flag `emitInline`,
which controlled whether template scoping information for the NgModule was
emitted directly into the compiled NgModule definition, or whether an
associated statement was generated which patched the information onto the
NgModule definition. Both options are useful in different contexts.

This commit changes this flag to an enum (and renames it), which allows for
a third option - do not emit any template scoping information. This option
is added to better represent the actual behavior of the Angular Linker,
which sometimes configures `compileNgModule` to use the side-effectful
statement generation but which does not actually emit such associated
statements. In other words, the linker effectively does not generate
scoping information for NgModules at all (in some configurations) and this
option more directly expresses that behavior.

This is a refactoring as no generated code is changed as a result of
introducing this flag, due to the linker's behavior of not emitting
associated statements.

JoostK commented on packages/compiler/src/render3/r3_module_compiler.ts:
```diff
@@ -13,6 +13,37 @@ import {Identifiers as R3} from './r3_identifiers';
+ */
+export enum R3SelectorScopeMode {
+  /**
+   * Omit the declarations inline into the module definition.
```
Omit -> Emit (same for the comment for the next variant)

alxhub committed:
fix(compiler-cli): full side-effectful registration of NgModules with ids

Angular contains an NgModule registry, which allows a user to declare
NgModules with string ids and retrieve them via those ids, using the
`getNgModuleById` API.

Previously, we attempted to structure this registration in a clever fashion
to allow for tree-shaking of registered NgModules (that is, those with ids).
This sort of worked due to the accidental alignment of behaviors from the
different tree-shakers involved. However, this trick relies on the
generation of `.ngfactory` files and how they're specifically processed in
various bundling scenarios. We intend to remove `.ngfactory` files, hence
we can no longer rely on them in this way.

The correct solution here is to recognize that `@NgModule({id})` is
inherently declaring a global side-effect, and such classes should not
really be eligible for tree-shaking in the first place. This commit removes
all the old registration machinery, and standardizes on generating a side-
effectful call to `registerNgModuleType` for NgModules that have ids.

There is some risk here that NgModules with unnecessary `id`s may not
tree-shake as a result of this change, whereas they would have in previous
circumstances. The fix here should be to remove the `id` if it's not needed.
Specifying an `id` is a request that the NgModule be retained regardless of
any other references, in case it is later looked up by string id.

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api, size-tracking

AndrewKushnir reviewed (commented):
Reviewed-for: size-tracking, public-api

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking, public-api

atscott reviewed (approved):
Reviewed-for: size-tracking, public-api

dylhunn reviewed (approved):
reviewed-for: public-api, size-tracking

jessicajaniuk commented:
Looks like legit test failures in the presubmit.

AndrewKushnir commented:
Note to the Caretaker: this PR will require an additional change in g3, please **include** [this change](https://critique.corp.google.com/cl/436314716) into the sync.

AndrewKushnir commented:
[Global Presubmit](https://test.corp.google.com/OCL:436364051:BASE:436364286:1647915894380:67281d72).

AndrewKushnir commented:
@alxhub FYI the CI was fixed in https://github.com/angular/angular/pull/45024/commits/4978d0d889c7ddbcef9a002355651a17b8ec1ed2, the TGP is fully "green" (with [an extra change](https://critique.corp.google.com/cl/436314716) that should be included in the sync). The PR should be ready for merge now.
